File: pset2-arrays/readability/readability.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <math.h>


// Run Coleman-Liau readability test on a text fragment
int main(void)
{
    // Store text fragment prompted by the user in 'text'
    string text = get_string("Text: ");

    // Initialize letter, word and sentence counts
    int letters = 0;
    int words = 1;
    int sentences = 0;

    // Iterate over every char in the string
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        // If text[i] is an alphabetic character, update letter count
        if (isalpha(text[i]))
        {
            letters++;
        }

        // If text[i] is a whitespace followed by a non-whitespace, update word count
        else if (isblank(text[i]) && !isblank(text[i + 1]))
        {
            words++;
        }

        // If text[i] is '.', '!' or '?', update sentence count
        else if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            sentences++;
        }
    }

    // Calculate Coleman-Liau index for text fragment
    float letter_average = (float)letters * 100 / (float)words;
    float sentence_average = (float)sentences * 100 / (float)words;
    int index = roundf(letter_average * 0.0588 - sentence_average * 0.296 - 15.8);

    // Print out grade level for text, according to Coleman-Liau formula
    if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
}

File: pset6-python/readability/readability.py
from cs50 import get_string


# Propmt user for text fragment
text = get_string("Text: ")

# Calculate number of letters, words and sentences in fragment
letters = sum(c.isalpha() for c in text)  # Returns sum of alphabetic characters in fragment
words = len(text.split())  # Splits fragment by whitespace and returns a list, then calculates its length
sentences = text.count(".") + text.count("?") + text.count("!")  # Counts instances of '.', '!' and '?', and returns total

# Calculate Coleman-Liau index for text fragment
letter_average = letters * 100 / words
sentence_average = sentences * 100 / words
index = round(letter_average * 0.0588 - sentence_average * 0.296 - 15.8)

# Print out grade level for text, according to Coleman-Liau formula
if index >= 16:
    print("Grade 16+")
elif index < 1:
    print("Before Grade 1")
else:
    print(f"Grade {index}")